### PR TITLE
Add support for managing Redis dnfmodule on EL 8+

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

File: spec/classes/redis_spec.rb [hunk 1]
```diff
@@ -81,6 +81,21 @@ class { 'redis::globals':
             end
           end
         end
+
+        describe 'with manage_dnf_module true', if: facts[:os]['family'] == 'RedHat' && facts[:os]['release']['major'].to_i > 7 do
```
Comment (ekohl) [inline spec/classes/redis_spec.rb:85]: I'm never a fan of `> x`
```suggestion
        describe 'with manage_dnf_module true', if: facts[:os]['family'] == 'RedHat' && facts[:os]['release']['major'].to_i >= 8 do
```

File: manifests/globals.pp [hunk 2]
```diff
@@ -2,10 +2,26 @@
...
 # @param scl
 #   Use a specific Software CoLlection on Red Hat based systems
+#
+# @param version
+#   The version of Redis to install and manage.
+#
+# @param manage_dnf_module
+#   Manage the DNF module. This only makes sense on distributions that use DNF
+#   package manager, such as EL8 or Fedora. It also requires Puppet 5.5.20+ or
+#   Puppet 6.15.0+ since they ship the dnfmodule provider.
```
Comment (ekohl) [inline manifests/globals.pp:12]: I'd be happy with changing `metadata.json` to 6.15.0+:
https://github.com/voxpupuli/puppet-redis/blob/38b0d2d4bb2abc0806e9f148ce385ae9a28deb38/metadata.json#L28-L33

File: manifests/globals.pp [hunk 3]
```diff
@@ -2,10 +2,26 @@
...
 class redis::globals (
   Optional[String] $scl = undef,
+  $version = '6',
```
Comment (ekohl) [inline manifests/globals.pp:15]: It may be confusing because the user can also pass in a version via `package_ensure`. There is also the `redis_server_version` fact (https://github.com/voxpupuli/puppet-redis/blob/master/lib/facter/redis_server_version.rb).

Should this be more focused and instead be `Optional[String[1]] $dnf_module_stream = undef` (and drop the `manage_dnf_module` parameter)?

File: manifests/dnfmodule.pp [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# @summary Manage the DNF module
+#
+# On EL8 and Fedora DNF can manage modules. This is a method of providing
+# multiple versions on the same OS. Only one DNF module can be active at the
```
Comment (ekohl) [inline manifests/dnfmodule.pp:4]:
```suggestion
# multiple versions on the same OS. Only one DNF module stream can be active at the
```

File: spec/classes/redis_spec.rb [hunk 5]
```diff
@@ -81,6 +81,21 @@ class { 'redis::globals':
...
+            class { 'redis::globals':
+              manage_dnf_module => true,
+              version           => '7',
+            }
+            PUPPET
+          end
+
+          it { is_expected.to compile.with_all_deps }
+
+          it { is_expected.to contain_package('redis').with_name('redis') }
```
Comment (ekohl) [inline spec/classes/redis_spec.rb:97]: Could you add `.that_requires('Package[redis dnf module]')` to this line?

File: spec/classes/redis_spec.rb [hunk 6]
```diff
@@ -81,6 +81,21 @@ class { 'redis::globals':
...
+            <<-PUPPET
+            class { 'redis::globals':
+              manage_dnf_module => true,
+              version           => '7',
+            }
+            PUPPET
+          end
+
+          it { is_expected.to compile.with_all_deps }
+
```
Comment (ekohl) [inline spec/classes/redis_spec.rb:96]: Could you add the expectation `contain_package('redis dnf module').with_ensure('7')`

File: manifests/globals.pp [hunk 7]
```diff
@@ -2,10 +2,26 @@
 #
 # @param scl
 #   Use a specific Software CoLlection on Red Hat based systems
```
Comment (ekohl) [inline manifests/globals.pp:4]:
```suggestion
#   Use a specific Software Collection on Red Hat 7 based systems
```

**Review by ekohl — CHANGES_REQUESTED** [review]
It would be good to verify the version really is 7 in an acceptance test. I recall calling facter on the command line while using custom Puppet facts was tricky so perhaps you can parse the output of `redis-server -v` like the fact does. It's rather trivial anyway and I suspect some of the code never triggers anymore.

**Comment by ehelms** [comment]
I can't seem to see what I am missing to make the requires relationship work in testing:

```
  1) redis on centos-8-x86_64 without parameters with manage_dnf_module true is expected to contain Package[redis dnf module] that comes before Package[redis]
     Failure/Error: it { is_expected.to contain_package('redis dnf module').with_ensure('6').that_comes_before('Package[redis]') }
       expected that the catalogue would contain Package[redis dnf module] with that comes before Package[redis]
     # ./spec/classes/redis_spec.rb:98:in `block (6 levels) in <top (required)>'

  2) redis on centos-9-x86_64 without parameters with manage_dnf_module true is expected to contain Package[redis dnf module] that comes before Package[redis]
     Failure/Error: it { is_expected.to contain_package('redis dnf module').with_ensure('6').that_comes_before('Package[redis]') }
       expected that the catalogue would contain Package[redis dnf module] with that comes before Package[redis]
     # ./spec/classes/redis_spec.rb:98:in `block (6 levels) in <top (required)>'

  3) redis on redhat-8-x86_64 without parameters with manage_dnf_module true is expected to contain Package[redis dnf module] that comes before Package[redis]
     Failure/Error: it { is_expected.to contain_package('redis dnf module').with_ensure('6').that_comes_before('Package[redis]') }
       expected that the catalogue would contain Package[redis dnf module] with that comes before Package[redis]
     # ./spec/classes/redis_spec.rb:98:in `block (6 levels) in <top (required)>'

  4) redis on redhat-9-x86_64 without parameters with manage_dnf_module true is expected to contain Package[redis dnf module] that comes before Package[redis]
     Failure/Error: it { is_expected.to contain_package('redis dnf module').with_ensure('6').that_comes_before('Package[redis]') }
       expected that the catalogue would contain Package[redis dnf module] with that comes before Package[redis]
```

File: spec/acceptance/suites/dnfmodule/redis_spec.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require 'spec_helper_acceptance'
+
+describe 'redis', if: %w[centos redhat].include?(os[:family]) && os[:release].to_i >= 8 do
+  redis_name = 'redis'
+
+  it 'runs successfully' do
```
Comment (ekohl) [inline spec/acceptance/suites/dnfmodule/redis_spec.rb:8]: Looks like CentOS 9 doesn't have a Redis 6 module (Redis 6.2.7 is a regular package).

File: spec/classes/redis_spec.rb [hunk 9]
```diff
@@ -81,6 +81,23 @@ class { 'redis::globals':
...
+              dnf_module_stream => '6',
+            }
+            class { 'redis':
+              manage_package => true,
+            }
+            PUPPET
+          end
+
+          it { is_expected.to compile.with_all_deps }
+          it { is_expected.to contain_package('redis dnf module').with_ensure('6').that_comes_before('Package[redis]') }
```
Comment (ekohl) [inline spec/classes/redis_spec.rb:98]: I suspect `Package<|tag == 'puppet-redis'|>` doesn't match the package. A simpler implementation may be to inline `redis::dnfmodule` in `redis::install` and use a regular relation rather than a collector. In that case you can also consider a regular class parameter in `init.pp` rather than one on globals. With SCL globals makes sense (since it affects paths, service names etc) while with DNF modules it's a lot simpler.

Comment (ehelms) [reply spec/classes/redis_spec.rb:98]: I took inspiration from the implementation [here[(https://github.com/puppetlabs/puppetlabs-postgresql/commit/8d0efec9fb5a5df42ed64375c71a11d8b6f90b4c) -- do you recall what inspired that or if your thinking around this has simply change to a more standard model?

Comment (ekohl) [reply spec/classes/redis_spec.rb:98]: Mostly because the PostgreSQL module is a lot more complex. For example, if you use the https://www.postgresql.org/ RPMs they're packaged very differently from Red Hat's RPMs. Like Debian's default packages, they are co-installable. That means the version comes back in paths like the data directory. As always, there's upsides and downsides to this.

Comment (ehelms) [reply spec/classes/redis_spec.rb:96]: I have changed it around to implement this strategy.

File: manifests/init.pp [hunk 10]
```diff
@@ -302,7 +302,10 @@
 # @param rdb_save_incremental_fsync
 #   When redis saves RDB file, if the following option is enabled
 #   the file will be fsync-ed every 32 MB of data generated.
-
+# @param dnf_module_stream
+#   Manage the DNF module and set the version. This only makes sense on distributions
+#   that use DNF package manager, such as EL8 or Fedora. It also requires Puppet 5.5.20+ or
+#   Puppet 6.15.0+ since they ship the dnfmodule provider.
```
Comment (ekohl) [inline manifests/init.pp:308]: Given the minimum version is specified in `metadata.json`, I'd leave it out here
```suggestion
#   that use DNF package manager, such as EL8 or Fedora.
```

File: spec/acceptance/suites/dnfmodule/redis_spec.rb [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+require 'spec_helper_acceptance'
+
+describe 'redis', if: %w[centos redhat].include?(os[:family]) && os[:release].to_i == 8 do
+  redis_name = 'redis'
+
+  it 'runs successfully' do
+    pp = <<-PUPPET
+      class { 'redis':
+        manage_repo       => true,
```
Comment (ekohl) [inline spec/acceptance/suites/dnfmodule/redis_spec.rb:11]: Why is this needed?

File: spec/acceptance/suites/dnfmodule/redis_spec.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+require 'spec_helper_acceptance'
+
+describe 'redis', if: %w[centos redhat].include?(os[:family]) && os[:release].to_i == 8 do
+  redis_name = 'redis'
+
+  it 'runs successfully' do
+    pp = <<-PUPPET
+      class { 'redis':
```
Comment (ekohl) [inline spec/acceptance/suites/dnfmodule/redis_spec.rb:10]: I suspect it runs after a test which has already installed Redis and then the ensure doesn't match. I suspect you can use `package_ensure` to upgrade.
```suggestion
      class { 'redis':
        package_ensure => 'latest',
```

**Review by ekohl — CHANGES_REQUESTED** [review]
The two unit test failures also happen on master so they're unrelated. Acceptance tests do appear related. See inline comment about `package_ensure`.

**Comment by ehelms** [comment]
To address the failing unit tests that are unrelated -- https://github.com/voxpupuli/puppet-redis/pull/452

**Review by bastelfreak — APPROVED** [review]

**Review by ekohl — APPROVED** [review]
